File: lessons/compiled_lang/go/src/f21.loop.go
// testbox: title="three-clause for as conditional loop"

package main

import (
	"bufio"
	"fmt"
	"os"
)

func main() {
	stdin := bufio.NewScanner(os.Stdin)

	// three-clause for with an empty post clause - answer is reassigned
	//  in the body instead
	for answer := ""; answer != "quit"; {
		fmt.Print("Enter your name (quit to Exit): ")
		stdin.Scan()
		answer = stdin.Text()

		if answer != "quit" {
			fmt.Println("Hello " + answer + "!")
		}
	}
}

File: lessons/compiled_lang/java/src/f21.loop.java
// testbox: title="for loop as conditional loop"
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;

class F21Loop {
    public static void main(String[] args) throws IOException {
        BufferedReader stdin = new BufferedReader(new InputStreamReader(System.in));

        // the for statement's own update clause is left empty - answer is
        //  reassigned in the body instead, so the loop's only real job
        //  here is testing the condition before each pass
        for (String answer = ""; !answer.equals("quit"); ) {
            System.out.print("Enter your name (quit to Exit): ");
            answer = stdin.readLine();

            if (!answer.equals("quit"))
                System.out.println("Hello " + answer + "!");
        }
    }
}
